### DISCUSSION TITLE
Transferring std::vector between apps

### DISCUSSION THREAD

**Discussion by sblondel** [idaholab/moose#18894] [Q&A General] [answered]
Hi,

I want to take advantage of the transfer system for an external app. I know how to transfer a scalar between apps but now I am interested in transferring a std::vector which size would be defined at run time. I looked into the MultiAppReporterTransfer but I am still not sure if it is the best approach. What would you recommend using?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Is it sized based on the number of elements?
In what format do you want this to be available on the MOOSE side? Variable? Vectorpostprocessor?

Guillaume

**Reply by sblondel** [reply]
@GiudGiud The size will be independent of the number of elements, but each mesh/grid point will be passing the same vector size. I think I would prefer to have it as a variable but as long as the externalApp can use it it doesn't matter to me.

Sophie

**Reply by GiudGiud** [reply]
So it's a vector for every element in the mesh? Like an array variable?

**Reply by sblondel** [reply]
I believe so.

**Reply by GiudGiud** [reply]
Great thanks! I dont think anyone has done that in particular. @aprilnovak has made regular variable transfers in Cardinal I believe. Could you please help out here? @sblondel you are at ORNL right? This will help with sharing that project.

@fdkong @permcody  do you know of any other external apps who worked with transfers in that direction before?

**Reply by sblondel** [reply]
I am official at University of Tennessee but have an collaborator account at ORNL. For transfers I am familiar with @fdkong work in the external_petsc_solver module.

**Reply by aprilnovak** [reply]
Hi @sblondel - we have a wrapping of the Nek CFD code in MOOSE that has transfer similar to this - but we don't use the Transfer system _directly_. Our general approach is to:

1. Create a mirror of the external App's mesh (on which the `std::vector` entries correspond to data on quadrature points, for instance) as a `MooseMesh`.
2. Within an `ExternalProblem` class, directly write from the `std::vector` into a `MooseVariable` on the mesh created in step (1).
3. Use MOOSE's usual transfer system to then get that variable to another application.

So in essence, we don't write any custom `Transfer`s - instead, we copy the external application's data onto a mesh that MOOSE understands, and then use any existing transfer to get data to another application. We chose not to go with the custom `Transfer` route because we would have had to implement all the logistics and parallelism of the transfers - like, finding the nearest node on a MooseMesh to a quadrature point in Nek, or doing a mesh interpolation/projection between a MOOSE mesh and a Nek mesh, etc. It was much simpler to just do a mesh + data copy and use MOOSE's transfers "out-of-the-box." 

What does your application look like?

**Reply by sblondel** [reply]
Hi @aprilnovak , your approach sounds good. To make sure I understand it: do you have multiple meshes for one app in this case (one for the physical domain and one for the std::vector entries)?

I'm trying to extend this application: https://github.com/SciDAC-MOOSE-Xolotl-coupling-group/coupling_xolotl/blob/master/src/problem/XolotlProblem.C .

**Reply by aprilnovak** [reply]
At all times, we have two meshes for Nek - one is the actual mesh that Nek uses for its CFD solve and on which the `std::vector` exists. The other is a "mirror" of that mesh in a `MooseMesh` format - no solves happen on this mesh - it only exists to receive the `std::vector` from Nek and put it into a `MooseVariable`.

I'm not familiar with your application, but it looks like you have a mesh that replicates the Xolotl domain, in `XolotlReflectedMesh`? I think our analogue is a `NekRSMesh`: https://cardinal.cels.anl.gov/source/mesh/NekRSMesh.html

To get the `std::vector` onto the mesh mirror, we have a `NekRSProblem` where we do this in `syncSolutions`. In `syncSolutions`, we call a function like this - `_nek_mesh` is a pointer to the mesh mirror, and the `value` we pass in is essentially the `std::vector` we extracted from Nek.

```
void
NekRSProblemBase::fillAuxVariable(const unsigned int var_number, const double * value)
{
  auto & solution = _aux->solution();
  auto sys_number = _aux->number();
  auto pid = _communicator.rank();

  for (unsigned int e = 0; e < _n_elems; e++)
  {
    auto elem_ptr = _nek_mesh->queryElemPtr(e);

    for (unsigned int n = 0; n < _n_vertices_per_elem; n++)
    {
      auto node_ptr = elem_ptr->node_ptr(n);

      // For each face vertex, we can only write into the MOOSE auxiliary fields if that
      // vertex is "owned" by the present MOOSE process.
      if (node_ptr->processor_id() == pid)
      {
        // Nek uses different node ordering than linMesh
        int node_index = _nek_mesh->nodeIndex(n);
        auto node_offset = e * _n_vertices_per_elem + node_index;

        // get the DOF for the auxiliary variable, then use it to set the value in the auxiliary system
        auto dof_idx = node_ptr->dof_number(sys_number, var_number, 0);
        solution.set(dof_idx, value[node_offset]);
      }
    }
  }

  solution.close();
}
```

Most importantly, in the `NekRSProblem::addExternalVariables` function, we add MOOSE variables to represent Nek's internal solution. It looks like your `XolotlProblem` doesn't have any MOOSE variables added like this. If you wanted to use this approach, you'd add an external variable to represent whatever Xolotl solves for (say, flux?). Then in `syncSolutions`, write into that flux variable. Then you should be able to apply any MOOSE transfer to that MOOSE-variable-representation-of-flux to get it somewhere else.

**Reply by sblondel** [reply]
Thank you so much for sharing this! Yes, `XolotlReflectedMesh` replicates the Xolotl domain. I'll try something similar to what you have and let you know how it went (it will probably take me a few days...).

**Reply by aprilnovak** [reply]
Are you on the MOOSE slack? Feel free to reach out if you have any questions, there's not a ton of examples of the external problem approach in the framework itself

**Reply by fdkong** [reply]
Thanks @aprilnovak for these detailed answers. It seems straightforward to implement in the coupling code. 

I am confused with one thing. If you have a std::vector, for example, size 3, at each quad point, and then what kind of variable can take data for all these points? A higher order elemental variable? As far as I know, the number of q points can be like 5 or more

**Reply by fdkong** [reply]
@sblondel If you want a chat on how to implement that in the coupling code, please let me know.

**Reply by aprilnovak** [reply]
Yes, in Nek our polynomial order is often 7 - so we have 8^3 quadrature points on a single Nek element. We build second-order libMesh elements (HEX27) and do an interpolation (with Vandemonde matrices) of Nek's 8^3 entries in the `std::vector` into the 27 nodal values on the HEX27 element (so the MOOSE variable is second lagrange). 

So the data isn't represented in exactly the same way. But if you really needed that, you could instead create (8^3)/8 HEX8 elements with nodes that correspond to the Nek quadrature points, and then write from the 8^3 Nek DOFs into linear lagrange variables on those HEX8 elements. We figured this was overkill when we mostly communicate heat fluxes, which by nature of the `DiffusionFluxAux` requires the flux in the master application to be constant monomial.

**Reply by fdkong** [reply]
Thanks for explanation. I understood now!

**Reply by sblondel** [reply]
After some more reading I think I understand the process. Let me know if anything is wrong in my simplified explanation adapted to the coupling code I use:
- in two 2D, the `XolotlReflectedMesh` uses QUAD4 elements, which means there are 4 nodes per element
- without changing the mesh definition I could pass a `std::vector` up to size 4 through this mesh at each Xolotl grid point (Xolotl uses finite differences so it does need a mesh internally, just a grid)
- to access the different nodes within the element I would use `to_node->dof_number(sync_rate.sys().number(), sync_rate.number(), i);` where `i` varies from 0 to 3 and `sync_rate` in the MOOSE variable that get transferred

Now, what if the `std::vector` I want to pass is much larger or multi-dimensional? Would it be possible to have an additional mesh in the app that reflects the structure of the `std::vector` (if the vector is of size M,N, the mesh would have M times N elements) and transfer data through that mesh within the same process if we are in parallel?

**Reply by aprilnovak** [reply]
Suppose that the `std::vector` has length `# elements * # nodes/element`. When we make the mesh mirror in the first place, we create it in the same "order" that the `std::vector` is defined. i.e. the first element we build on the mesh mirror has nodes given by `my_vector[0][0 to 3]`, and the second element we build on the mesh mirror has nodes given by `my_vector[1][0 to 3]` and so on. Then when we get to actually writing data into a MooseVariable, we loop in the same order and index into `my_vector` the same way as we loop.

```
for (int e = 0; e < number_of_elements; ++e)
{
  auto elem_ptr = mesh.query_elem_ptr(e);
  for (int n = 0; n < number_of_nodes_per_elem; ++n)
  {
    auto node_ptr = elem_ptr->node_ptr(n);
    auto dof = node_ptr->dof_number(sys_number, _variable_number_that_im_setting, 0);
    solution.set(dof, my_vector[e][n]);
  }
}
```

Does that capture your question?

**Reply by sblondel** [reply]
So the mesh mirror you have is only used to reflect the `std::vector` you want to pass. It potentially reflects your spacial domain if your `std::vector` also represents the spacial domain in some way. 
In my application I would want two things: a mesh reflecting the spacial domain from Xolotl (by default `XolotlReflectedMesh`), and at each element of that mesh be able to store a `std::vector` of arbitrary size (not related to the number of elements, defined at runtime, can reach hundreds or thousands) so that it can be transferred. Is there a limit to the number of nodes per element that can be defined?
Sorry it's taking me many steps to understand, I am still novice with MOOSE.

**Reply by aprilnovak** [reply]
Correct. We don't do any solves on the mesh mirror - it represents our spatial domain, and we have several `std::vector`s that we map onto that space that represent quantities like velocity, pressure, and temperature. 

What does your `std::vector` represent? If it has multiple "fields" of a solution (like for fluids, we have multiple solution variables for velocity, pressure, and temperature), then the approach I'd use is to write to multiple different MooseVariables, one variable for each solution field from Xolotl. But I'm not sure I understand fully - if `std::vector` is unrelated to the number of elements, what reason is there to try and represent it as a MooseVariable on a mesh?

**Reply by sblondel** [reply]
To put it simply it would represent the reaction part of the RHS entries the sub app needs to solve (Xolotl is a cluster dynamic code). For this application I want to take advantage of the MOOSE multiApp system (transfer, time step control) but am not interested in anything else on the MOOSE side (and we already did the work of wrapping Xolotl in MOOSE to couple to phase field). All the solves would be done by the wrapped Xolotl apps. Here is a schematic view:
![xolotl-hierarchy](https://user-images.githubusercontent.com/12696451/135301460-2153fd03-74b9-48dc-909b-4a98b4a5e64e.png)
The master app knows about the whole cluster network we need to solve on but doesn't solve anything, and each sub app only knows about sub parts of the cluster network and solves the diffusion reaction equations. The reaction equations would be updated every cycle to take into account the change of concentrations from the other sub apps, computed by the master app. The idea is to decouple the reaction equations as the rates can be greatly different depending on the type of clusters reacting. In the end there would be two transfers: each sub app sends their solution vectors to the master app containing the concentrations; the master app sends the updated RHS coefficients to each sub app.
I wanted to represent this vector on the mesh because it is the only way I knew MOOSE could exchange data between apps.

**Reply by aprilnovak** [reply]
Ok that makes sense now - if the data doesn't make sense to represent on a mesh, then maybe my suggestions aren't the right way (at least for the RHS coefficients you're transferring). I think @aeslaughter's suggestion might be the way to go then, but I don't know much about the Reporter system myself.

**Reply by sblondel** [reply]
The more I think about it and the more I realize I probably don't need a transfer on a mesh... I could have my external app owning the multiple instances of Xolotl and do the transfer directly between instances at the end of each MOOSE time step without using a MOOSE variable.
@aprilnovak Thank you so much for taking the time to explain everything!

**Comment by aeslaughter** [comment]
The Reporter system should be able to handle this, perhaps @zachmprince can provide some additional insight into this problem.

**Comment by zachmprince** [comment]
Hi @sblondel. Just give an overview of the reporter system. It has the same high-level functionality as Postprocessors and VectorPostprocessors, except the data can be of arbitrary type. The data created with reporters is available throughout the entire application and can be referenced by other objects. Or it can be transferred between sub-applications using MultiAppReporterTransfer, as long as the type is copy-assignable.

So from my possibly flawed understanding of what you need. You could maybe create a Reporter object that converts the data from the Xototl run into an `std::vector`. You can then create empty `std::vector` reporter values in the main and other sub-apps using ConstantReporter which MultiAppReporterTransfer can transfer the data into. The objects in the sub-app running the other Xototl instances can then reference this transferred data for their execution.

I hope this makes sense 😬

**Reply by sblondel** [reply]
Thank you very much for the overview. I'm still trying to decide what is the best approach for my case, I'll let you know if I run into issues with the `Reporter` and `MultiAppReporterTransfer`.
